scripts/artifacts.py

```diff
@@ -0,0 +1,263 @@
+@click.option(
+    "--out",
+    type=click.Path(file_okay=True, dir_okay=False, writable=True),
+    required=True,
+)
+@click.option("--limit", type=int, default=10)
+def plot_stats(out, ref, limit):
+    print("Fetching latest commits.", file=sys.stderr)
+    infos = fetch_commit_infos(ref, limit=limit)
+    stats = list(map(mk_commit_stats, tqdm(infos, desc="Fetching artifacts")))
+    stats.reverse()
+
+    fig, axs = plt.subplots(2)
+    fig.set_size_inches(14, 10)
+
+    xticks = list(range(0, len(stats)))
+    xlabels = [i.info.ref for i in stats]
+
+    # gas costs
+    axs[0].set_title("Gas Costs")
+    for key in set(it.chain(*[i.gas_costs for i in stats])):
+        axs[0].plot(xticks, [i.gas_costs.get(key) for i in stats], label=key)
+
+    # entrypoint sizes
+    axs[1].set_title("Entrypoint Sizes")
+    for key in set(it.chain(*[i.entrypoint_sizes for i in stats])):
+        axs[1].plot(xticks, [i.entrypoint_sizes.get(key) for i in stats], label=key)
+
+    # save the graph
+    for ax in axs:
+        ax.set_xticks(xticks)
+        ax.set_xticklabels(xlabels, rotation=65)
+        ax.legend(bbox_to_anchor=(1.05, 1), loc="upper left")
+
+    plt.tight_layout()
+    plt.savefig(out)
+
+
+@cli.command()
+@click.option("--previous", required=True)
+@click.option("--next", required=True)
+def compare_stats(previous, next):
+    previous_stats = mk_commit_stats(fetch_commit_info(previous))
+    next_stats = mk_commit_stats(fetch_commit_info(next))
+
+    print_diffs(
+        mk_diffs(previous_stats.gas_costs, next_stats.gas_costs),
+        title="Gas costs",
+        header_previous=previous_stats.info.ref,
+        header_next=next_stats.info.ref,
+    )
+
+    print_diffs(
+        mk_diffs(previous_stats.entrypoint_sizes, next_stats.entrypoint_sizes),
+        title="Entrypoint sizes",
+        header_previous=previous_stats.info.ref,
+        header_next=next_stats.info.ref,
+    )
+
+
+# From the given $ref (usually a commit hash or a branch name), fetch the latest $limit commits
+# (from children to parents) and their workflow ids.
+def fetch_commit_infos(ref, *, limit):
+    query = """
+    query($ref:String!, $limit:Int!) {
+      repository(owner: "tezos-checker", name: "checker") {
+        object(expression: $ref) {
+          ... on Commit {
+            history(first: $limit) {
+              nodes {
+                message
+                oid
+                statusCheckRollup {
+                  contexts(first: 10) {
+                    edges {
+                      node {
+                        ... on CheckRun {
+                          checkSuite {
+                            workflowRun {
+                              databaseId
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
```

Comment:
Some serious brackets here :sweat_smile:

Reply:
Yeah... A possible alternative is to use a GraphQL client library for Python that'd maybe allow us to form the query with first class values. But we only have a single query, so just hardcoding the request body also works just fine.

Reply:
No complaints here, I just thought it was a bit funny :smile: